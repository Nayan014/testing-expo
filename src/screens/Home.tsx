import React, { useState } from "react";
import {
  Button,
  FlatList,
  Modal,
  StyleSheet,
  Text,
  TextInput,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from "react-native";

interface toDoItem {
  task: string;
  isComplected: boolean;
}

const Home: React.FC = () => {
  const [modalVisible, setModalVisible] = useState<boolean>(false);
  const [task, setTask] = useState("");
  const [editTask, setEditTask] = useState<number | null>(null);
  const [taskList, setTaskList] = useState<toDoItem[]>([]);

  const taskItem = ({ item, index }: { item: toDoItem; index: number }) => {
    return (
      <TouchableOpacity
        key={`${index}${item.task}`}
        onPress={() => {
          setEditTask(index);
          setTask(item.task);
          setModalVisible(true);
        }}
      >
        <View style={{ padding: 14 }}>
          <Text>
            {item.task.slice(0, 1).toUpperCase() + item.task.slice(1)}
          </Text>
        </View>
      </TouchableOpacity>
    );
  };

  const addEditTask = () => {
    if (!task) return;
    if (editTask !== null) {
      setTaskList((preState) => {
        preState[editTask].task = task;
        return preState;
      });
      setEditTask(null);
    } else {
      setTaskList((preState) => {
        preState.push({
          task,
          isComplected: false,
        });
        return preState;
      });
    }

    setTask("");
    setModalVisible(!modalVisible);
  };

  const modalView = () => (
    <Modal
      animationType="slide"
      transparent={true}
      visible={modalVisible}
      onRequestClose={() => {
        setTask("");
        setModalVisible(!modalVisible);
      }}
    >
      <TouchableWithoutFeedback
        onPress={() => {
          setTask("");
          setModalVisible(false);
        }}
      >
        <View style={styles.modalContainer}>
          <View style={styles.modalView}>
            <View style={{ width: "100%" }}>
              <TextInput
                style={styles.input}
                placeholder="Enter Task"
                value={task}
                onChangeText={(text) => setTask(text)}
              />
              <View style={{ marginTop: 18 }}>
                <Button
                  title={editTask ? "Edit Task" : "Add Task"}
                  onPress={addEditTask}
                />
              </View>
            </View>
          </View>
        </View>
      </TouchableWithoutFeedback>
    </Modal>
  );

  const addButton = () => (
    <View style={{ position: "absolute", bottom: 18, right: 18 }}>
      <View style={styles.addContainer}>
        <TouchableOpacity onPress={() => setModalVisible(true)}>
          <Text style={styles.addText}>+</Text>
        </TouchableOpacity>
      </View>
    </View>
  );

  return (
    <View style={{ marginTop: 12, flex: 1 }}>
      <FlatList data={taskList} renderItem={taskItem} />
      {modalView()}
      {!modalVisible && addButton()}
    </View>
  );
};

export default Home;

const styles = StyleSheet.create({
  addContainer: {
    height: 70,
    width: 70,
    backgroundColor: "red",
    borderRadius: 100,
    borderWidth: 1,
    borderColor: "red",
    alignItems: "center",
    justifyContent: "center",
  },
  addText: {
    fontSize: 48,
    color: "white",
    marginBottom: 4,
  },
  modalContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
  },
  modalView: {
    margin: 20,
    backgroundColor: "white",
    borderRadius: 8,
    padding: 35,
    alignItems: "center",
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
    width: "80%",
  },
  input: {
    borderColor: "gray",
    borderWidth: 0.5,
    borderRadius: 5,
  },
});
