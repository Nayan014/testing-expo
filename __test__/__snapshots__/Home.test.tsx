import { fireEvent, render } from '@testing-library/react-native';
import Home from '../../src/screens/Home';
import { debug } from '@testing-library/react-native/build/helpers/debug';

describe('Home Screen', () => {
  it('snapshot testing', () => {
    const tree = render(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should open model when click on +', () => {
    const { getByText, getByPlaceholderText } = render(<Home />);

    const addButton = getByText('+');
    fireEvent.press(addButton);
    const addTaskText = getByPlaceholderText('Enter Task');

    expect(addTaskText).toBeTruthy();
  });

  it('able to add task ', () => {
    const { getByText, getByPlaceholderText } = render(<Home />);
    const addButton = getByText('+');
    fireEvent.press(addButton);
    const addTaskInput = getByPlaceholderText('Enter Task');
    const submitTask = getByText('Add Task');
    expect(addTaskInput).toBeTruthy();
    fireEvent.changeText(addTaskInput, 'New Task');
    fireEvent.press(submitTask);
    expect(getByText('New Task')).toBeTruthy();
  });

  it('should able to edit task', () => {
    const { getByText, getByPlaceholderText, debug, getByTestId } = render(
      <Home />
    );
    const addButton = getByText('+');
    fireEvent.press(addButton);
    fireEvent.changeText(getByPlaceholderText('Enter Task'), 'New Task');
    fireEvent.press(getByText('Add Task'));

    expect(getByText('New Task')).toBeTruthy();
    fireEvent.press(getByText('New Task'));

    // Change text in the input
    fireEvent.changeText(getByTestId('taskInput'), 'Edited Task');

    /*  await waitFor(() => {
      expect(getByTestId('taskInput').props.value).toBe('Edited Task');
    }); */

    debug();

    //fireEvent.press(getByText('Edit Task'));

    //expect(getByText('Edited Task')).toBeTruthy();
  });

  it('outside click should close the model', () => {
    const { getByText, getByTestId, queryByPlaceholderText } = render(<Home />);
    const addButton = getByText('+');
    fireEvent.press(addButton);
    const background = getByTestId('modalBackground');
    fireEvent.press(background);
    expect(queryByPlaceholderText('Enter Task')).toBeNull();
  });
});
