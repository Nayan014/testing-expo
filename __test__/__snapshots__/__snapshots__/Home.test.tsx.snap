// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen snapshot testing 1`] = `
<View
  style={
    {
      "flex": 1,
      "marginTop": 12,
    }
  }
>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      {
        "bottom": 18,
        "position": "absolute",
        "right": 18,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "red",
          "borderColor": "red",
          "borderRadius": 100,
          "borderWidth": 1,
          "height": 70,
          "justifyContent": "center",
          "width": 70,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 48,
              "marginBottom": 4,
            }
          }
        >
          +
        </Text>
      </View>
    </View>
  </View>
</View>
`;
